### PR TITLE
busybox compatibility changes

### PR DESCRIPTION
**Pull request by dermotbradley**
This patch replaces long options for cmp, getopt, and truncate with
short options that are supported by both Busybox's implementation
and also the full implementations.

### CODE REVIEW

**Comment by dermotbradley** [comment]
@paride could you review this please?

**Comment by paride** [comment]
@dermotbradley Hi, sorry for not getting to this yet, the reason is that I have something slightly different in mind. I'd like to prepare a branch that runs the tests in CI against busybox. You could then fix the test failures, and as the final result will be better busybox support with less chance of regressions, as we're going to keep the busybox testing in CI.

I already tried locally (via `busybox --install -s` and then altering `$PATH` accordingly), and the approach seems to work.

If you agree with this plan I can push a commit adding busybox CI to your branch, then you can work on the test fixes. What do you think?

**Comment by dermotbradley** [comment]
> I'd like to prepare a branch that runs the tests in CI against busybox. You could then fix the test failures, and as the final result will be better busybox support with less chance of regressions, as we're going to keep the busybox testing in CI.

I'm happy with that in principle.

This MR is that last of the patches I (currently) have for cloud-utils. I have plans to raise some minor shellcheck-related MRs in the future but no timescale for those.

I was hoping after this MR was merged to request that a new cloud-utils release (0.33) be made as its been 1.5 years since the last one. I was hoping for this occur in time for me to package up a new cloud-utils release in time for the next Alpine release in a couple of weeks time.

With your proposal above I guess a new release is not going to happen in time for that...

**Comment by dermotbradley** [comment]
> I pushed some extra commits to your branch, could you please review those? All the changes should be straightforward, maybe with the exception of the `grep /proc/mounts` one in `mount-image-callback`.

They look fine. I hadn't checked on Busybox compatibility for the tests, only for the run-time scripts, as generally speaking the Alpine package builders have the full coreutils/util-linux tools installed.

One minor question regarding the use in test-growpart and test-growpart-lvm of:

```
! has_cmd udevadm || udevadm settle
```

versus what is used elsewhere:

```
has_cmd udevadm && udevadm settle
```

Any particular reason for using the "!" and the OR?

File: bin/mount-image-callback
Comment (smoser) [inline bin/mount-image-callback:127]: This will have false positives for two reasons:
- it will match mountpoints that start with the same path. /tmp/foo will match /tmp/foobar while /tmp/foobar is not under /tmp/foo. 
- presence of the string in fields other than mountpoint (device or options)

**Comment by paride** [comment]
> Any particular reason for using the "!" and the OR?

I recall having to use the OR logic to make some script play well with `set -e` (which the tests use, while the tools do not use), but now I can't remember the details. I know of the exception for the left hand operand of an `&&` or `||`, which is working as expected. I'll switch those lines to the more straightforward `&&` logic, thanks for bringing it up.

**Review by smoser — COMMENTED** [review]
over all it feels like a step backwards to me to use short flags.  I understand why it is necessary here, but `--size=100G` is so much better than `-s 100G` for 2 reasons:
 a.) it is clear that this is 1 argument
 b.) i dont have to go looking at a man page to figure out what '-s' means.  Does that mean "silent" or is the 100G a value and it mean 'size'?

File: bin/mount-image-callback
Comment (paride) [reply bin/mount-image-callback:127]: Thanks Scott for reviewing, that's the bit when I was most hoping to get feedback. Do you think that's actually a concern, given that `TEMP_D` has a random chunk?
```
TEMP_D=$(mktemp -d "${TMPDIR:-/tmp}/${0##*/}.XXXXXX")
```
What if we make the random part longer, just to be safe?

**Comment by paride** [comment]
@smoser I agree, actually I learned actively using long flags in scripts from your review comments. This was discussed briefly with @holmanb and @blackboxsw and we agreed that after all it's worth switching to short flags for wider compatibility.

**Comment by paride** [comment]
As the tests run with `set -e` (and I'm not going to try to drop it now) I'd leave the negative `! has_cmd udevadm` logic in place. Positive `&&` logic should work because of `set -e` exceptions, but I prefer to avoid having a command group exiting nonzero.

**Comment by smoser** [comment]
> As the tests run with `set -e` (and I'm not going to try to drop it now) I'd leave the negative `! has_cmd udevadm` logic in place. Positive `&&` logic should work because of `set -e` exceptions, but I prefer to avoid having a command group exiting nonzero.

"because of `set -e` exceptions" ? what does that  mean.
set -e is a trainwreck it can't be sanely used. https://hackmd.io/IoVN08RrQV6Hyb6ztY4kwg

**Comment by paride** [comment]
> "because of `set -e` exceptions" ? what does that mean. set -e is a trainwreck it can't be sanely used. [here](https://hackmd.io/IoVN08RrQV6Hyb6ztY4kwg) is an article I wrote when I finally gave up on it full stop.

It means that in `set -e` mode (quoting bash(1)):

> The shell does not exit if the command failed [...] is part of any command executed in a `&&` or `||` list except the command following the final `&&` or `||`

so basically `false && true` won't make the shell exit, even if the final RC is non zero. It's this kind of exceptions or special rules that make `set -e` especially bad (see "story time" https://mywiki.wooledge.org/BashFAQ/105).

(Personally I'm not *that* against `set -e`, I think it's useful in very simple scripts, but I agree it's not a good idea in more complex ones.)

File: bin/mount-image-callback
Comment (paride) [reply bin/mount-image-callback]: There is also `findmnt --submounts --mountpoint` we could use. I'll check tomorrow what we can do to improve on the current approach.

Comment (paride) [reply bin/mount-image-callback]: I think `grep "^[^ ]\+ ${TEMP_D}[/ ]" /proc/mounts` will do. I can see only two ways it could fails:
1. Field separator not a single space. The format of `/proc/mounts` is the one described in fstab(5), which says "Fields on each line are separated by tabs or spaces", so that's not forbidden, but I don't think the kernel is ever going to do that.
2. First field has spaces. I wonder if there's anything forbidding this, and I wonder how other tools parsing the file would behave.

Comment (paride) [reply bin/mount-image-callback]: I created a block device with a whitespace (`mknod "/dev/bad bad" b 8 17`), mounted it and `/proc/mounts` has:
```
/dev/bad\040bad /mnt vfat ...
```
so we're safe from point 2.

Comment (paride) [reply bin/mount-image-callback]: I added the check you suggested, but also left the `grep /proc/mounts` check, which should be solid and I think worth keeping as we're `rm -Rf`ing stuff.

File: bin/mount-image-callback
Comment (smoser) [inline bin/mount-image-callback:125]: i may be wrong, but i think that UMOUNTS may *not* be all under TMPDIR.
if you told it where to mount (--mountpiont), then the mounts would be in a different location.

What i was suggsting was :+1: 
```bash
if [ -d "${TEMP_D}" ]; then
    umount_r "${TEMP_D}"
fi
```

the thing that we'd have to change is to make umount_r not fail if it did not unmount anything.

File: bin/mount-image-callback
Comment (smoser) [inline bin/mount-image-callback:127]: i think this logic is broken, no?
if TEMP_D is empty or it is not a directory, then you remove it?

File: bin/mount-image-callback
Comment (paride) [reply bin/mount-image-callback:125]: Maybe not all UMOUNTS are under TMPDIR, but all the TMPDIR mounts (if any) should be in UMOUNTS, and this should be enough for us, I think?

File: bin/mount-image-callback
Comment (paride) [reply bin/mount-image-callback:127]: Thanks for spotting this, my fault when converting the check from "inline" to an `if`.

File: bin/mount-image-callback
Comment (smoser) [reply bin/mount-image-callback:125]: well you could have mounts that were done not by this program.

    mount-image-callback my.img --chroot -- mount -t /dev /dev  /dev

Comment (paride) [reply bin/mount-image-callback:125]: I still don't get why we need an extra `umount_r`. TLDR: can you please provide a reproducer?

---

Assuming that you mean something like:
```
mount-image-callback my.img mchroot -- mount -t devtmpfs /dev /dev
```
from the script's point of view (i.e. from outside the chroot) the `/dev` mount will be under the mountpoint, which is under `TEMP_D` (if `--mountpoint` is not specified). The mountpoint is added to `UMOUNTS` and therefore umounted *recursively*. I verified this by doing
```
mount-image-callback autopkgtest-bionic-amd64.img mchroot -- sh -c 'mount -t devtmpfs /dev /dev ; sleep 60'
```
and in mounts I have:
```
/dev/nbd0p1 on /tmp/mount-image-callback.VQs86dyvlH/mp type ext4 (rw,relatime)
/dev on /tmp/mount-image-callback.VQs86dyvlH/mp/dev type devtmpfs (rw,relatime,size=12187540k,nr_inodes=3046885,mode=75mount_callback_umount5,inode64)
```
and everything is cleaned up correctly after `sleep` is done.

I also tried to manually mount something under the mountpoint from a different shell (mount not done by the script, not done from the chroot). It got unmounted just fine.

If I `cd mountpoint/dev` to interfere with the unmount process, I get:
```
$ mount-image-callback -vvv autopkgtest-images/autopkgtest-bionic-amd64.img mchroot -- sh -c 'mount -t devtmpfs /dev /dev ; sleep 30'
pid for /dev/nbd1 is 816180
connected /home/paride/ubuntu/autopkgtest-images/autopkgtest-bionic-amd64.img (qcow2) auto to /dev/nbd1. waiting for device.
partitioned disk.
mounted /dev/nbd1p1 via qemu-nbd /dev/nbd1 at /tmp/mount-image-callback.CMaV7qBW6m/mp
invoking: MOUNTPOINT=/tmp/mount-image-callback.CMaV7qBW6m/mp mchroot sh -c mount -t devtmpfs /dev /dev ; sleep 30
cmd returned 0. unmounting /tmp/mount-image-callback.CMaV7qBW6m/mp
umount: /tmp/mount-image-callback.CMaV7qBW6m/mp/dev: target is busy.
WARNING: unmounting filesystem at /tmp/mount-image-callback.CMaV7qBW6m/mp failed!
failed umount /home/paride/ubuntu/autopkgtest-images/autopkgtest-bionic-amd64.img
disconnecting /dev/nbd1
umounts: /tmp/mount-image-callback.CMaV7qBW6m/mp
umount: /tmp/mount-image-callback.CMaV7qBW6m/mp/dev: target is busy.
WARNING: unmounting filesystem at /tmp/mount-image-callback.CMaV7qBW6m/mp failed!
failed cleaning up mounts
```
as expected. Note that it tried twice: once from `mount_callback_umount()` and once from `cleanup()`.

If (e.g. because of a bug) the script silently failed to umount something, the `grep -q` will detect it and prevent the the `rm -Rf` when `TEMP_D` is in an unknown state.

If you still believe there's an issue here, could you please come up with a more concrete reproducer?

Also: do you think the current logic is in any significant way weaker than what we had before (`rm --one-file-system`)? If you don't think so, but still believe you spotted an issue which was already present, we could work on it outside of this "busybox compat" PR.

(Your idea of having `umount_r` return 2 in the "nothing to umount" case has some issues, but IMHO not worth discussing if we agree that we don't need it after all.)

Comment (smoser) [reply bin/mount-image-callback:125]: I think your right. 

I guess the one chance for failure is if umount_r falls. In that case,  then I guys there is no reason to rm at all. But check the umount return and avoid rm if it failed?

Comment (paride) [reply bin/mount-image-callback:125]: The check is there:
```
do_umounts "${UMOUNTS[@]}" ||
    { error "failed cleaning up mounts"; return 1; }
```
If that `return 1` is reached the only cleanup operation that is skipped is the `rm -Rf $TEMP_D`.

Comment (smoser) [reply bin/mount-image-callback:125]: You’re right again. I was remembering how it was abs didn’t realize you’d changed that. 

So now I think you can drop the grep below on added line 128.

Comment (paride) [reply bin/mount-image-callback:125]: We can, but I'm thinking of leaving it as an extra safety measure before the `rm -Rf`, just in case e.g. a future bug in `umount_r` makes it fail umount but not return failure, or UMOUNTS gets corrupted.

(Side note: TIL that util-linux umount has `--recursive`, but we're doing these changes for BusyBox compatibility, so that's not an option.)

Comment (paride) [reply bin/mount-image-callback:125]: I added a comment about the `grep`, however I have no strong feelings about it, if you think it's useless or confusing I'll just drop it.

Comment (smoser) [reply bin/mount-image-callback:125]: if you have the grep to "protect", then youll never find out about the bug in umount_r.
I'm personally not a fan of extra code that "shouldnt ever be needed".
but I'll leave that up to you.

Comment (paride) [reply bin/mount-image-callback:125]: Well if the grep fails an error is printed (= bugs in `umount_r` are found), and I was hoping that the latest version of the grep check was free from false positives (see my comment where I mention the fstab(5) format). For reference:

```
if ! grep -q "^[^ ]\+ ${TEMP_D}[/ ]" /proc/mounts; then
    rm -Rf "$TEMP_D" || error "removal of temp dir failed!"
else
    error "not removing temp dir (stale mounts)!"
fi
```

However I'm now convinced that the grep is unnecessary (it's especially your comment on "extra comment that shouldn't ever be needed" that made me reconsider). I now dropped it.

File: bin/mount-image-callback
Comment (paride) [reply bin/mount-image-callback]: You are right, I moved it back after the umounts.

**Review by smoser — APPROVED** [review]
thanks for your patience. all my suggests are addressed.

**Pull request merged**
